core/src/com/unciv/ui/pickerscreens/ImprovementPickerScreen.kt

```diff
@@ -72,13 +86,16 @@ class ImprovementPickerScreen(
             // canBuildImprovement() would allow e.g. great improvements thus we need to exclude them - except cancel
             if (improvement.turnsToBuild == 0 && improvement.name != Constants.cancelImprovementOrder) continue
             if (improvement.name == tileInfo.improvement) continue // also checked by canImprovementBeBuiltHere, but after more expensive tests
-            if (!tileInfo.canBuildImprovement(improvement, currentPlayerCiv)) {
-                // if there is an improvement that could remove that terrain
-                if (!tileInfoNoLast.canBuildImprovement(improvement, currentPlayerCiv)) continue
-                suggestRemoval = true
-            }
             if (!unit.canBuildImprovement(improvement)) continue
 
+            var unbuildableBecause = tileInfo.getImprovementBuildingProblems(improvement, currentPlayerCiv).toSet()
+            if (!canReport(unbuildableBecause)) {
+                // Try after pretending to have removed the top terrain layer.
+                unbuildableBecause = tileInfoNoLast.getImprovementBuildingProblems(improvement, currentPlayerCiv).toSet()
```

Comment:
tileInfoWithoutLastTerrain maybe?

Reply:
This variable existed before, I didn't add it. Renamed anyway.

Reply:
This variable existed before, I didn't add it. Renamed anyway.